pallets/gear/src/migrations.rs

```diff
@@ -0,0 +1,311 @@
+// This file is part of Gear.
+//
+// Copyright (C) 2024 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{Config, Pallet, QueueOf};
+use common::{
+    event::{MessageWokenSystemReason, SystemReason},
+    storage::Queue,
+    Origin,
+};
+use frame_support::{
+    traits::{Get, GetStorageVersion, OnRuntimeUpgrade},
+    weights::Weight,
+};
+use gear_core::ids::MessageId;
+use sp_std::{marker::PhantomData, vec::Vec};
+
+#[cfg(feature = "try-runtime")]
+use {
+    common::storage::IterableMap,
+    parity_scale_codec::{Decode, Encode},
+    sp_runtime::TryRuntimeError,
+};
+
+pub struct MigrateWaitingInitList<T>(PhantomData<T>);
+
+impl<T> OnRuntimeUpgrade for MigrateWaitingInitList<T>
+where
+    T: Config + pallet_gear_program::Config,
+    T::AccountId: Origin,
+{
+    fn on_runtime_upgrade() -> Weight {
+        let current = pallet_gear_program::Pallet::<T>::current_storage_version();
+        let onchain = pallet_gear_program::Pallet::<T>::on_chain_storage_version();
+
+        log::info!(
+            "🚚 Running migration with current storage version {current:?} / onchain {onchain:?}"
+        );
+
+        // 1 read for the on-chain storage version
+        let mut weight = T::DbWeight::get().reads(1);
+
+        if current == 4 && onchain == 3 {
+            waiting_init_list::WaitingInitStorage::<T>::translate(
+                |program_id, messages: Vec<MessageId>| {
+                    // read and remove an element
+                    weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+
+                    for message_id in messages {
+                        if let Some(dispatch) = Pallet::<T>::wake_dispatch(
```

Comment:
I think we should remove old tasks on migration also, because this can lead to panics in runtime: 
https://github.com/gear-tech/gear/blob/6316fc9fd1f8e52c3f383d93b16a45bd2415811b/pallets/gear/src/manager/task.rs#L112-L120

Reply:
Totally agree. You have the key for pallet-scheduler from waitlist - it keeps deadline in there next to message itself.

Reply:
`wake_dispatch` automatically removes task

Reply:
Yep, see now. Could you please append the test for that